crates/wascc-provider/src/states/starting.rs

```diff
@@ -0,0 +1,174 @@
+            let container_port = c_port.container_port;
+            if let Some(host_port) = c_port.host_port {
+                let host_port: u16 = u16::try_from(host_port)?;
+                let mut lock = port_map.lock().await;
+                if !lock.contains_key(&host_port) {
+                    port_assigned = host_port;
+                    lock.insert(port_assigned, key_from_pod(pod));
+                } else {
+                    error!(
+                        "Failed to assign hostport {}, because it's taken",
+                        &host_port
+                    );
+                    return Err(anyhow::anyhow!("Port {} is currently in use", &host_port));
+                }
+            } else if container_port >= 0 && container_port <= 65536 {
+                port_assigned = find_available_port(&port_map, pod).await?;
+            }
+        }
+    }
+    Ok(port_assigned)
+}
+
+async fn start_container(
+    pod_state: &mut PodState,
+    container: &Container,
+    pod: &Pod,
+    port_assigned: u16,
+) -> anyhow::Result<ContainerHandle<ActorHandle, LogHandleFactory>> {
+    let env =
+        <WasccProvider as Provider>::env_vars(&container, &pod, &pod_state.shared.client).await;
+    let volume_bindings: Vec<VolumeBinding> =
+        if let Some(volume_mounts) = container.volume_mounts().as_ref() {
+            volume_mounts
+                .iter()
+                .map(|vm| -> anyhow::Result<VolumeBinding> {
+                    // Check the volume exists first
+                    let vol = pod_state.run_context.volumes.get(&vm.name).ok_or_else(|| {
+                        anyhow::anyhow!(
+                            "no volume with the name of {} found for container {}",
+                            vm.name,
+                            container.name()
+                        )
+                    })?;
+                    // We can safely assume that this should be valid UTF-8 because it would have
+                    // been validated by the k8s API
+                    Ok(VolumeBinding {
+                        name: vm.name.clone(),
+                        host_path: vol.deref().clone(),
+                    })
+                })
+                .collect::<anyhow::Result<_>>()?
+        } else {
+            vec![]
+        };
+
+    debug!("Starting container {} on thread", container.name());
+
+    let module_data = pod_state
+        .run_context
+        .modules
+        .remove(container.name())
+        .expect("FATAL ERROR: module map not properly populated");
+    let lp = pod_state.shared.log_path.clone();
+    let host = pod_state.shared.host.clone();
+    tokio::task::spawn_blocking(move || {
+        wascc_run_http(host, module_data, env, volume_bindings, &lp, port_assigned)
+    })
+    .await?
+}
+
+state!(
+    /// The Kubelet is starting the Pod.
+    Starting,
+    PodState,
+    {
+        info!("Starting containers for pod {:?}", pod.name());
```

Comment:
I've got a concern around discoverability with blocks like this.  For me at least, I'm getting no code completion in VS Code, and there's nothing that tells me what parameters I have available to me.  Is it possible to make these, say, a lambda, or a function name?

Reply:
There is a second macro option that accepts a function (although now that I look at it, it needs to be updated to accept the `PodState`) I'm not sure if this will fix the VS Code issue. But in general I agree, this requires developers to understand a lot of magic behind the macro. I'm wondering if the macro is really that useful, the trait itself is not especially verbose. We might consider reserving the macro for the `graph` default state graph implementation that @thomastaylor312 has proposed.

Reply:
Yeah, we do have a macro that just takes a function path and then calls it so that should solve the auto completion thing. We could just expect people to implement the `State` trait by hand now that it is a bit simpler

Reply:
I'm going to ignore this for now since the use of the macro or the trait is entirely up to the user.